Search index in PrestaShop (#1763)
PrestaShop/docs: development/products-lifecycle/search-index.md

---
title: Products search index
weight: 42
useMermaid: true
---

# Products search index

In PrestaShop, products are natively indexed for search in database, by keywords. 

When a query is made in a search bar, it is sanitized, splitted in words, and queries are made against `ps_search_word`. 

> Then, hits are retrieved from `ps_search_index`, to retrieve matching product ids, and then a weighting and a sort is made to return relevant products. 

Reviewer: ```suggestion
Then, hits are retrieved from `ps_search_index`, to retrieve matching product ids, and then a weighting and a sort are made to return relevant products. 
```

> Then, hits are retrieved from `ps_search_index`, to retrieve matching product ids, and then a weighting and a sort is made to return relevant products. 

Reviewer 2: In PrestaShop, product search functionality relies on keyword-based indexing. Each search query entered in the search bar undergoes sanitization and is split into individual words. These words are then matched against the `ps_search_word` table. Matching product IDs are retrieved from the `ps_search_index`, followed by the process of weighting and sorting to deliver the most relevant product results.

> In PrestaShop, products are natively indexed for search in database, by keywords. 
> 
> When a query is made in a search bar, it is sanitized, splitted in words, and queries are made against `ps_search_word`. 
> 
> Then, hits are retrieved from `ps_search_index`, to retrieve matching product ids, and then a weighting and a sort is made to return relevant products. 

Reviewer 2: ```suggestion
In PrestaShop, product search functionality relies on keyword-based indexing. Each search query entered in the search bar undergoes sanitization and is split into individual words. These words are then matched against the `ps_search_word` table. Matching product IDs are retrieved from the `ps_search_index`, followed by the process of weighting and sorting to deliver the most relevant product results.
```

<div class="mermaid">
flowchart TB
    id1[Search Query] --> id2[Sanitize,\nremove unwanted words,\nsplit by words]
    id2 --> id3[Retrieve `id_word` from `ps_search_word`]
    id3 --> id4[Retrieve `id_product` and `weight` from `ps_search_index`]
    id4 --> id5[Weight results to return most relevant products]
</div>

## Search index structure

<div class="mermaid">
classDiagram
    ps_product <-- ps_search_index
    ps_search_word <-- ps_search_index
    ps_product : int product_id
    ps_product : ...
    class ps_search_index{
        int id_product
        int id_word
        int weight
    }
    class ps_search_word{
        int id_word
        int id_shop
        int id_lang
        varchar word    
    }
</div>

## Search index lifecycle

> Several actions can trigger a reindex of a Product or of the complete catalog in the database: 

Reviewer 2: ```suggestion
There are several actions that can trigger a reindex of a product or the complete catalog in the database:
```

| pa_reference | `PS_SEARCH_WEIGHT_REF` | 10 | Combination reference |
| supplier_reference | `PS_SEARCH_WEIGHT_REF` | 10 | Supplier reference
| pa_supplier_reference | `PS_SEARCH_WEIGHT_REF` | 10 | Combination supplier reference |
| ean13 | `PS_SEARCH_WEIGHT_REF` | 10 | Product EAN13 |
| pa_ean13 | `PS_SEARCH_WEIGHT_REF` | 10 | Combination EAN13 |
| isbn | `PS_SEARCH_WEIGHT_REF` | 10 | Product ISBN |
| pa_isbn | `PS_SEARCH_WEIGHT_REF` | 10 | Combination ISBN |
| upc | `PS_SEARCH_WEIGHT_REF` | 10 | Product UPC |
| pa_upc | `PS_SEARCH_WEIGHT_REF` | 10 | Combination UPC |
| mpn | `PS_SEARCH_WEIGHT_REF` | 10 | Product MPN |
| pa_mpn | `PS_SEARCH_WEIGHT_REF` | 10 | Combination MPN |
| description_short | `PS_SEARCH_WEIGHT_SHORTDESC` | 1 | Product short description |
| description | `PS_SEARCH_WEIGHT_DESC` | 1 | Product description |
| cname | `PS_SEARCH_WEIGHT_CNAME` | 3 | Category name |
| mname | `PS_SEARCH_WEIGHT_MNAME` | 3 | Manufacturer name |
| tags | `PS_SEARCH_WEIGHT_TAG` | 4 | Product tags |
| attributes | `PS_SEARCH_WEIGHT_ATTRIBUTE` | 2 | Combinations |
| features | `PS_SEARCH_WEIGHT_FEATURE` | 2 | Product features |

## Trigger a Search Index refresh by cron

> To trigger a Search Index refresh by cron, craft an url to be called with GET method, to the Back Office, to the Admin controller **AdminSearch**. 

Reviewer 2: ```suggestion
To trigger a search index refresh via cron, create a GET request URL to the Back Office Admin controller, **AdminSearch**.
```

| Param | Value | Description |
| --- | --- | --- |
| action | searchCron | |
| ajax | 1 | |

> | full | 1 | If 1, it will rebuild the full index. If 0 or omitted, it will build only missing products |

Reviewer 2: ```suggestion
| full | 1 | If 1, it will rebuild the full index. If 0 or omitted, it will index only missing products |
```

| token | **tokenValue** | |

> To create the **tokenValue**, you need to retrieve the `_COOKIE_KEY_` constant from the Configuration. 
> 
> Then, your token is created by extracting a substring from this constant. 
> 
> ```php
> $tokenValue = substr(
>     _COOKIE_KEY_,
>     AdminSearchController::TOKEN_CHECK_START_POS,
>     AdminSearchController::TOKEN_CHECK_LENGTH
> );
> ```
> 
> If correctly crafted, your URL should look like: 
> 
> ```
> https://domain.tld/admin-xxx/index.php?controller=AdminSearch&action=searchCron&ajax=1&full=1&token=xxxxxxxx
> ```

Reviewer 2: ```suggestion
```

{{% notice note %}}

> You can also find this URL already generated in `Back Office > Shop Parameters > Search > Indexing`

Reviewer 2: ```suggestion
You can find indexation URL in `Back Office > Shop Parameters > Search > Indexing`
```

{{% /notice %}}

> Then, your URL can be used by cURL in a cron: 

Reviewer 2: ```suggestion
Instead of manually running the script, you can use an indexation URL with cURL in a crontab.
```
